File: sdk/csharp/src/Providers/WebMailTemp.cs
using System;
using System.Collections.Generic;
using System.Net;
using System.Text;
using System.Text.Json.Nodes;

namespace XxxXTeam.TempMail.Providers;

/// <summary>
/// WebMailTemp 渠道 — https://webmailtemp.com。
/// 创建：GET /api/create 返回 email、username、ttl，并下发 session cookie。
///       token 序列化保存 username 与 cookie 头串（wmt1: + base64）。
/// 收信：GET /api/check/{username}，携带 cookie，返回 emails 数组。
/// </summary>
public static class WebMailTemp
{
    private const string BaseUrl = "https://webmailtemp.com";
    private const string TokPrefix = "wmt1:";
    private static readonly Dictionary<string, string> Headers = new()
    {
        ["Accept"] = "application/json",
        ["User-Agent"] = "Mozilla/5.0",
    };

    private static string EncodeToken(string username, string cookie)
    {
        var raw = Encoding.UTF8.GetBytes(Json.Serialize(new Dictionary<string, object?> { ["u"] = username, ["c"] = cookie }));
        return TokPrefix + Convert.ToBase64String(raw);
    }

    private static (string Username, string Cookie) DecodeToken(string token)
    {
        if (!token.StartsWith(TokPrefix, StringComparison.Ordinal)) throw new Exception("webmailtemp: invalid token");
        var o = Json.Parse(Encoding.UTF8.GetString(Convert.FromBase64String(token[TokPrefix.Length..]))) as JsonObject;
        var username = Json.Str(o, "u").Trim();
        var cookie = Json.Str(o, "c").Trim();
        if (string.IsNullOrEmpty(username) || string.IsNullOrEmpty(cookie)) throw new Exception("webmailtemp: invalid token data");
        return (username, cookie);
    }

    private static Dictionary<string, object?> Flatten(JsonObject raw, string recipient) => new()
    {
        ["id"] = Json.Str(raw, "id"),
        ["from"] = Json.Str(raw, "from"),
        ["to"] = recipient,
        ["subject"] = Json.Str(raw, "subject"),
        ["text"] = Json.Str(raw, "body"),
        ["html"] = Json.Str(raw, "html"),
        ["date"] = !string.IsNullOrEmpty(Json.Str(raw, "received_at")) ? Json.Str(raw, "received_at") : Json.Str(raw, "timestamp"),
        ["attachments"] = raw["attachments"] is JsonArray a ? Json.ToRaw(a) : null,
    };

    /// <summary>创建临时邮箱</summary>
    public static EmailInfo Generate()
    {
        var resp = Http.Get($"{BaseUrl}/api/create", new Dictionary<string, string>(Headers));
        resp.EnsureSuccess();
        var data = Json.Parse(resp.Body) as JsonObject;
        var email = Json.Str(data, "email").Trim();
        var username = Json.Str(data, "username").Trim();
        var cookieHeader = FirstCookie(resp);
        var success = (data?["success"] as JsonValue)?.GetValue<bool>() ?? false;
        if (!success || string.IsNullOrEmpty(email) || !email.Contains('@')
            || string.IsNullOrEmpty(username) || string.IsNullOrEmpty(cookieHeader))
            throw new Exception("webmailtemp: invalid create response");

        long? expires = null;
        if (data!.TryGetPropertyValue("ttl", out var ttlNode) && ttlNode is JsonValue tv
            && tv.TryGetValue<double>(out var ttl) && ttl > 0)
            expires = (long)((DateTimeOffset.UtcNow.ToUnixTimeMilliseconds() / 1000.0 + ttl) * 1000);

        return new EmailInfo("webmailtemp", email, EncodeToken(username, cookieHeader), expires);
    }

    /// <summary>获取邮件列表</summary>
    public static List<Email> GetEmails(string? token, string email)
    {
        var address = (email ?? "").Trim();
        if (string.IsNullOrEmpty(address)) throw new Exception("webmailtemp: empty email");
        var (username, cookie) = DecodeToken(token ?? "");

        var h = new Dictionary<string, string>(Headers) { ["Cookie"] = cookie };
        var resp = Http.Get($"{BaseUrl}/api/check/{WebUtility.UrlEncode(username)}", h);
        resp.EnsureSuccess();
        var rows = (Json.Parse(resp.Body) as JsonObject)?["emails"] as JsonArray;
        var result = new List<Email>();
        if (rows is null) return result;

        foreach (var r in rows)
            if (r is JsonObject row)
                result.Add(Normalize.NormalizeEmail(Flatten(row, address), address));
        return result;
    }

    // 取响应中第一条 Set-Cookie 的 name=value（对齐 Python 的 next(iter(cookies))）
    private static string FirstCookie(HttpResult resp)
    {
        if (resp.SetCookies.Count == 0) return "";
        var raw = resp.SetCookies[0];
        var semi = raw.IndexOf(';');
        return (semi < 0 ? raw : raw[..semi]).Trim();
    }
}

File: sdk/php/src/Providers/WebMailTemp.php
<?php

declare(strict_types=1);

namespace ChanhanzhanX\TempMail\Providers;

use ChanhanzhanX\TempMail\Email;
use ChanhanzhanX\TempMail\EmailInfo;
use ChanhanzhanX\TempMail\HttpClient;
use ChanhanzhanX\TempMail\Normalize;
use Psr\Http\Message\ResponseInterface;

/**
 * webmailtemp.com 渠道实现
 *
 * API: https://webmailtemp.com
 * 流程: GET /api/create 创建邮箱（email + username，并保存 Set-Cookie）→
 *       GET /api/check/{username} 读信（携带 Cookie）。
 * token 序列化保存 username 与 Cookie 头（base64）。
 */
final class WebMailTemp
{
    private const CHANNEL = 'webmailtemp';
    private const BASE_URL = 'https://webmailtemp.com';
    private const TOK_PREFIX = 'wmt1:';

    /** @var array<string,string> */
    private const HEADERS = [
        'Accept' => 'application/json',
        'User-Agent' => 'Mozilla/5.0',
    ];

    public static function generate(): EmailInfo
    {
        $resp = HttpClient::get(self::BASE_URL . '/api/create', self::HEADERS);
        $data = HttpClient::json($resp);
        $email = trim((string) ($data['email'] ?? ''));
        $username = trim((string) ($data['username'] ?? ''));
        $cookieHeader = self::firstCookie($resp);
        if (
            empty($data['success'])
            || $email === ''
            || !str_contains($email, '@')
            || $username === ''
            || $cookieHeader === ''
        ) {
            throw new \RuntimeException('webmailtemp: 创建邮箱响应无效');
        }
        $ttl = $data['ttl'] ?? null;
        $expiresAt = (is_numeric($ttl) && (float) $ttl > 0)
            ? (int) ((time() + (float) $ttl) * 1000)
            : null;
        return new EmailInfo(self::CHANNEL, $email, self::encodeToken($username, $cookieHeader), expiresAt: $expiresAt);
    }

    /**
     * @return Email[]
     */
    public static function getEmails(string $email, ?string $token): array
    {
        $address = trim($email);
        if ($address === '') {
            throw new \InvalidArgumentException('webmailtemp: 邮箱地址不能为空');
        }
        $session = self::decodeToken((string) $token);

        $resp = HttpClient::get(
            self::BASE_URL . '/api/check/' . rawurlencode($session['username']),
            self::HEADERS + ['Cookie' => $session['cookie']],
        );
        $data = HttpClient::json($resp);
        $rows = $data['emails'] ?? null;
        if (!is_array($rows)) {
            return [];
        }

        $out = [];
        foreach ($rows as $row) {
            if (is_array($row)) {
                $out[] = Normalize::email(self::flatten($row, $address), $address);
            }
        }
        return $out;
    }

    /**
     * 提取响应中的首个 Set-Cookie（name=value）。
     */
    private static function firstCookie(ResponseInterface $resp): string
    {
        foreach ($resp->getHeader('Set-Cookie') as $sc) {
            $first = trim(explode(';', $sc, 2)[0]);
            if ($first !== '' && str_contains($first, '=')) {
                return $first;
            }
        }
        return '';
    }

    private static function encodeToken(string $username, string $cookie): string
    {
        $raw = (string) json_encode(['u' => $username, 'c' => $cookie], JSON_UNESCAPED_SLASHES);
        return self::TOK_PREFIX . base64_encode($raw);
    }

    /**
     * @return array{username:string,cookie:string}
     */
    private static function decodeToken(string $token): array
    {
        if (!str_starts_with($token, self::TOK_PREFIX)) {
            throw new \InvalidArgumentException('webmailtemp: 无效的 token');
        }
        $decoded = base64_decode(substr($token, strlen(self::TOK_PREFIX)), true);
        if ($decoded === false) {
            throw new \InvalidArgumentException('webmailtemp: 无效的 token');
        }
        $data = json_decode($decoded, true);
        $username = is_array($data) ? trim((string) ($data['u'] ?? '')) : '';
        $cookie = is_array($data) ? trim((string) ($data['c'] ?? '')) : '';
        if ($username === '' || $cookie === '') {
            throw new \InvalidArgumentException('webmailtemp: token 数据不完整');
        }
        return ['username' => $username, 'cookie' => $cookie];
    }

    /**
     * @param array<mixed> $raw
     * @return array<string,mixed>
     */
    private static function flatten(array $raw, string $recipient): array
    {
        return [
            'id' => $raw['id'] ?? '',
            'from' => $raw['from'] ?? '',
            'to' => $recipient,
            'subject' => $raw['subject'] ?? '',
            'text' => $raw['body'] ?? '',
            'html' => $raw['html'] ?? '',
            'date' => ($raw['received_at'] ?? '') ?: ($raw['timestamp'] ?? ''),
            'attachments' => $raw['attachments'] ?? [],
        ];
    }
}
